Initial pass at allowing single-shard transactions

jasonmp85 opened the pull request:
Presently requires that all modifications within that transaction go to a single shard. Still going to figure out what to do about queries that use other executor types.

Silently accepts `SAVEPOINT` commands, but rejects `ROLLBACK`s by `ERROR`ing, aborting the entire transaction (and sending `ABORT` to remotes).

Fixes #505

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -66,12 +76,57 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
-		PreventTransactionChain(topLevel, "distributed commands");
 		eflags |= EXEC_FLAG_SKIP_TRIGGERS;
+
+		if (IsTransactionBlock())
```
Hm. For modifications happening from a multi-statement command, or from within a function, this won't work reliably. I think we should also enter that mode when context != PROCESS_UTILITY_TOPLEVEL or something similar.

jasonmp85 replied:
Chatted with @anarazel on chat about this. He said he misremembered and that detecting multi-statement/function contexts for non-utility statements has no trivial solution, so we're skipping this for now:

> Yea, there's no easy method. So let's skip that for now.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -66,12 +76,57 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
+
+		if (IsTransactionBlock())
+		{
+			if (shardForTransaction == INVALID_SHARD_ID)
```
Can we move this logic into a separate function, we're 6 blocks indented below...

jasonmp85 replied:
Sure, I'll refactor.

anarazel commented:
> Presently requires that all modifications within that transaction go to a single shard. Still going to figure out what to do about queries that use other executor types.

That's pretty much why I suggested using connection establishment for error checking. Makes it easy to centrally forbid dangerous situations.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -39,17 +40,26 @@
 bool AllModificationsCommutative = false;
 
+static uint64 shardForTransaction = INVALID_SHARD_ID;
+static bool callbackInstalled = false;
```
should probably reference 'xact' or something in th evarname, there could be many differen types of callbacks.

jasonmp85 replied:
I'll look into whether this state variable could be removed somehow. If not, I'll add `xact` to the name.

jasonmp85 replied:
Removing in favor of using the `shmem` callbacks to install the transaction callbacks.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -620,3 +756,109 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+{
+	if (txnStarted && event == SUBXACT_EVENT_ABORT_SUB)
+	{
+		UnregisterSubXactCallback(RouterSubtransactionCallback, NULL);
```
You're installing that one conditional on ```callbackInstalled```, won't uninstalling it here cause issues?

jasonmp85 replied:
Not entirely sure what this means, but I'm abandoning the logic in this callback anyhow so no worries.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -39,17 +40,26 @@
 
+static uint64 shardForTransaction = INVALID_SHARD_ID;
+static bool callbackInstalled = false;
+static bool txnStarted = false;
```
I think should make clearer that it's about router having started remote xacts.

jasonmp85 replied:
I think I meant it as _local transaction is in progress_, but I'll consider some sort of name change.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -620,3 +756,109 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+	{
+		UnregisterSubXactCallback(RouterSubtransactionCallback, NULL);
+
+		AbortOutOfAnyTransaction();
```
I don't think it's ok to do that here. Why do you want to fully abort the transaction here? That'll reset the current transaction status! I.e. we'll be in TRANS_DEFAULT after it. The ereport() below will leave the status in DEFAULT, because there's nothing going on. Which means that a later ROLLBACK/COMMIT from the clientside will not do anything meaningful.   Given that we want the rest of AbortSubTransaction() to run, we shouldn't ereport() there in the first place!  I think the easiest thing for now would be to just have a flag transactionIsAborted or something, that we check before committing.

jasonmp85 replied:
> I think the easiest thing for now would be to just have a flag transactionIsAborted or something, that we check before committing.

I thought about it but was already somewhat embarrassed by the amount of state I've added to this flow.

Shouldn't `ROLLBACK` be an _error_? We don't support it. It should cause the client to throw an exception. This invocation was the sole thing I could come up with that exits out of the entire transaction and sends and `ERROR` to the client. I don't think the user should be permitted to proceed past a `ROLLBACK` that we don't honor.

anarazel replied:
> > +	txnStarted = false;
> > +}
> > +
> > +
> > +/*
> > + * pgfdw_subxact_callback --- cleanup at subtransaction end.
> > + */
> > +static void
> > +RouterSubtransactionCallback(SubXactEvent event, SubTransactionId subId,
> > +							 SubTransactionId parentSubid, void *arg)
> > +{
> > +	if (txnStarted && event == SUBXACT_EVENT_ABORT_SUB)
> > +	{
> > +		UnregisterSubXactCallback(RouterSubtransactionCallback, NULL);
> > +
> > +		AbortOutOfAnyTransaction();
> 
> > I think the easiest thing for now would be to just have a flag transactionIsAborted or something, that we check before committing.
> 
> I thought about it but was already somewhat embarrassed by the amount of state I've added to this flow.
> 
> Shouldn't `ROLLBACK` be an _error_? We don't support it.

you mean ROLLBACK TO SAVEPOINT, right? ROLLBACK itself should be
supported.


> It should cause the client to throw an exception. This invocation was
> the sole thing I could come up with that exits out of the entire
> transaction and sends and `ERROR` to the client. I don't think the
> user should be permitted to proceed past a `ROLLBACK` that we don't
> honor.

The point is that the current implementation is broken (transaction
state lost, cleanups missed), and there's no trivial way to implement it
correctly (as there's no POST_ABORT_SUB hook).  The cases where it's
likely to be encountered is from within a UDF's exception handling, so
we can't even intercept the ROLLBACK TO SAVEPOINT utility command.
Having state that handles this with an error at the end of transaction
seems ok enough to me.

- Andres

jasonmp85 replied:
Adding another `static` file variable to remember that a subxact abort was attempted, then erroring in the main xact callback.

Still kind of weird, since the user will be able to deduce that the `ROLLBACK` wasn't actually applied using e.g. `SELECT`s within the remaining transaction. No biggie since it won't be applied anyway.

jasonmp85 commented:
I've addressed all your comments. Anything else I should start improving here?

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -53,6 +61,11 @@ static bool SendQueryInSingleRowMode(PGconn *connection, char *query);
 static bool StoreQueryResult(MaterialState *routerState, PGconn *connection,
 							 TupleDesc tupleDescriptor, int64 *rows);
 static bool ConsumeQueryResult(PGconn *connection, int64 *rows);
+static void RegisterRouterExecutorXactCallbacks();
```
Without void in the params, that's an old style function declaration (i.e. taking arbitrary parameters, without checking).

jasonmp85 replied:
I don't know how this got back in. A compiler complained at me about it at one point and I fixed it. Must have been during rebase; sorry.

jasonmp85 replied:
Fixing.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -109,6 +132,56 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
+
+		if (txnParticipant->connection != NULL)
+		{
+			PGresult *result = PQexec(connection, "BEGIN");
```
Hm. If there's a number of placements, the cost of the roundtrip could be mitigated by sending the BEGIN asynchronously to each, and only waiting at the end. The current way scales linearly with the number of placements, the other near constant (if looking at RTT).  Not sure if it has to be done now, but it seems worth mentioning.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -344,8 +425,20 @@ ExecuteTaskAndStoreResults(QueryDesc *queryDesc, Task *task,
+			PG_CATCH();
+			{
+				/* this placement is already aborted; remove it from participant list */
+				participantList = list_delete_ptr(participantList, taskPlacement);
```
list_delete_ptr is O(n). Harmless enough here, but in future you might want to look at list_delete_cell instead (or use ilists.h style lists).

jasonmp85 replied:
Yeah, I checked out the implementation before using the list things I'm going here (in particular, was worried about quadratic performance when using deletion in loops), but in typical use the replication factor is probably two or three (does anyone use four?), so the optimization seemed premature.

I'll check out your suggestions.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -751,3 +902,135 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+static void
+RegisterRouterExecutorXactCallbacks()
+{
+	RegisterXactCallback(RouterTransactionCallback, NULL);
```
Need to make sure whether work already has performed here. On windows / EXEC_BACKEND environments the startup hooks get called on each backend startup (since there's no fork to copy state).

anarazel replied:
Hm. On the other hand, that's probably what we want. Hm.

jasonmp85 replied:
> the startup hooks get called on each backend startup (since there's no fork to copy state)

Isn't that what I want here? I need these to run on every backend, which is why they're in the `shmem_startup_hook`, which I thought ran during backend initialization. Did I do this incorrectly?

jasonmp85 replied:
😦 Hoped this implementation (with the `shmem_startup_hook`) avoided the need for that state variable. So you want me to add it back for an `Assert` or something?

anarazel commented on src/include/distributed/multi_router_executor.h:
```diff
@@ -9,16 +9,26 @@
+{
+	ShardPlacement placement;
+	PGconn *connection;
+} TxnParticipant;
```
Maybe I'm pedantic, but imo type definitions should be before variable declarations (after all, the latter might need the former).

jasonmp85 replied:
Oops. I'll check other headers, but this is likely just an honest mistake.

anarazel commented on src/include/distributed/multi_router_executor.h:
```diff
@@ -9,16 +9,26 @@
 #ifndef MULTI_ROUTER_EXECUTOR_H_
 #define MULTI_ROUTER_EXECUTOR_H_
 
+#include "libpq-fe.h"
```
micro-sized-isssue: personally I prefer forward declaring PGconn...

jasonmp85 replied:
Do you have some sort of rule for when you want to do this?

anarazel replied:
Whenever the include leads to including a significant amount of code into places that don't need the information. E.g. there's no need for multi_executor.c to know about libpq and such.

jasonmp85 replied:
Ack. Fixing. Was curious because e.g. `include-what-you-use` has some functionality to suggest forward-declaration and I haven't been using it for some time but wondered if it might have made the same suggestion.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -751,3 +902,135 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+
+
+/*
+ * pgfdw_subxact_callback --- cleanup at subtransaction end.
```
Comment indicates the wrong function name, and the wrong purpose ;)

jasonmp85 replied:
Yeah, I need a general function comment pass on this before merging.

anarazel replied:
.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -344,8 +425,20 @@ ExecuteTaskAndStoreResults(QueryDesc *queryDesc, Task *task,
-			queryOK = StoreQueryResult(routerState, connection, tupleDescriptor,
-									   &currentAffectedTupleCount);
+			/* use try/catch to maintain txn participant list during fail-fast errors */
+			PG_TRY();
```
Hm. That pretty much assumes the error is a forwarded remote error. What if some memory allocation or such failed instead? We'll leave the connection open afaics. I'm a bit inclined towards stopping purging such connections instead.

anarazel commented:
Generally looks pretty good.  But I do think we should change the transaction management to be connection based, instead of shardid, as mentioned before. The current method will be more restrictive and harder to extend.

jasonmp85 commented:
> I do think we should change the transaction management to be connection based, instead of shardid, as mentioned before

Is that e.g. something to be done as part of this PR, or another task (I can do either).

anarazel commented:
@jasonmp85 I think it should be done in this PR.

jasonmp85 commented:
@anarazel I've reworked the main logic to do the connection-based thing we discussed.

I'm now letting the "default" logic just update placements as they fail; if the whole transaction rolls back, the operation marking things as unhealthy will roll back as well.

The one edge case is if we've got two shards (A and B), each on the same two nodes (1 and 2). If we have a workload like:

```sql
INSERT INTO table VALUES ('goes to A');
INSERT INTO table VALUES ('goes to B');
```

This would be permitted, because the first modification set the "participant" set of connections to nodes 1 and 2 and the second modification does not expand that list.

However, let's say node 2 has an issue with the first `INSERT`, which goes through on node 1. The placement of shard A on node 2 is immediately marked as unhealthy, meaning it does not participate in the second `INSERT`; one might assume we could still use the placement of shard B on node 2, but remember the connection is shared for all operations to node 2.

So if a connection is bad, we don't recycle it, instead just returning NULL, which will result in the B placement also being marked inactive, because node 2 cannot participate further in the transaction.

Finally, we also have to keep track of all the shards used on each connection in case the final COMMIT fails. If other nodes succeed but some fail, the failing nodes must have the shards that were modified on them marked as unhealthy.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -21,13 +21,16 @@
 #include "executor/executor.h"
 #include "nodes/pg_list.h"
 #include "utils/builtins.h"
+#include "storage/ipc.h"
```
shouldn't that be up one higher?

jasonmp85 replied:
I'll run a sort pass on this before flattening.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -69,12 +87,63 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
-		PreventTransactionChain(topLevel, "distributed commands");
 		eflags |= EXEC_FLAG_SKIP_TRIGGERS;
+
+		if (IsTransactionBlock() && xactParticipantHash == NULL)
```
Hm. Not that I have a better idea, but this implies that we'll not enter this block when called from a function :(

anarazel replied:
Hm. What about checking QueryDesc->dest, and treating everything but DestRemote, DestRemoteExecute as commands needing a transaction? I think that should do what we want.

jasonmp85 replied:
> Hm. Not that I have a better idea, but this implies that we'll not enter this block when called from within a function :(

I agree with this (and didn't have a better idea, other than some ugly hacks that would work but be slow).

> Hm. What about checking `QueryDesc->dest`, and treating everything but `DestRemote`, `DestRemoteExecute` as commands needing a transaction? I think that should do what we want.

_This_ I do not understand. Can you clarify what you're proposing? It sounds clean and like it could handle the other "multi-statements-inside-transactions" cases (functions and multi-command strings), but I don't know what `QueryDesc->dest` has to do with those cases. Quick explanation?

anarazel replied:
> > Hm. What about checking `QueryDesc->dest`, and treating everything but `DestRemote`, `DestRemoteExecute` as commands needing a transaction? I think that should do what we want.
> 
> _This_ I do not understand. Can you clarify what you're proposing? It sounds clean and like it could handle the other "multi-statements-inside-transactions" cases (functions and multi-command strings), but I don't know what `QueryDesc->dest` has to do with those cases. Quick explanation?

->dest is the destination where the results of a query a directed to. As
it happens, '*Remote' are the ones used when sending the results of a
query directly to the client (via the socket). The rest are used when
the results instead are received by other methods - and those pretty
much happen only when being called via a function, inside COPY, and such
stuff.  If we use a transaction for everything but *Remote, we have the
'function' use-case nailed, without incuring the overhead for
transaction-less commands from the client directly.

Makes more sense now?

jasonmp85 replied:
OK, so this'll solve functions (though is that by convention or definition? Could there be a PL which evades this detection?), but not multi-command strings from a client (which I'm inclined to ignore), right?

jasonmp85 replied:
(From Hangouts: PLs will inevitably use the non-remote destination, otherwise they won't get the result, so this solution covers all functions. Forget about multi-command strings.)

anarazel replied:
@jasonmp85 Could you update with that adjustment?

jasonmp85 replied:
Tried it out and we realized this shouldn't be done lightly (it affects other customers' use of `INSERT` within their own functions, many of which presently hit multiple shards within a single function).

jasonmp85 replied:
(i.e. it's going in a separate issue)

anarazel replied:
Maybe add a short comment about functions?

jasonmp85 replied:
Done.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -69,12 +87,63 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
+			}
+
+			MemoryContextSwitchTo(oldContext);
+		}
```
I think it'd be good to put that in a separate function.

jasonmp85 replied:
:grin: It was, but during the rewrite, this logic got shorter, so I again inlined it. I'll un-inline it (again).

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -751,3 +859,275 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+
+/* TODO: write comment */
+void
+InstallRouterExecutorShmemHook()
```
(void) in param list.

jasonmp85 replied:
Fixing.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -751,3 +859,275 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+
+/* TODO: Write comment */
+static void
+RegisterRouterExecutorXactCallbacks()
```
(void) in param list.

jasonmp85 replied:
Fixing.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -751,3 +859,275 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+AllocateUint64(uint64 value)
+{
+	uint64 *allocatedValue = (uint64 *) palloc0(sizeof(uint64));
+	Assert(sizeof(uint64) >= 8);
```
Huh?

jasonmp85 replied:
Copy-pasta. Not going to change, since I don't remember writing it. I'll get rid of my copy-pasted function in my clean-up commit.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -751,3 +859,275 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+	XactParticipantEntry *participantEntry = NULL;
+	bool entryFound = false;
+
+	if (xactParticipantHash == NULL)
```
Deserves a comment.

jasonmp85 replied:
OK. The presence of `xactParticipatingHash` has basically replaced the need for an "am I in a transaction?" `static bool`, so I'll note it where it occurs.

jasonmp85 replied:
(Or maybe I'll also explain the above in a short paragraph where I declare `xactParticipantHash`)

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -751,3 +859,275 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+
+
+static PGconn *
+GetConnectionForPlacement(ShardPlacement *placement)
```
Hm. Don't we have to use a function similar to this one, but without it calling RecordParticipatingShardId, for select queries too?

jasonmp85 replied:
Gah, I missed that side-effect. Yeah, we only need to mark "dirtied" shards as participating in the transaction for health reasons. I'll add a `readOnly` parameter here that'll skip the `RecordParticipatingShardId` if true.

anarazel commented on src/test/regress/sql/multi_modifying_xacts.sql:
```diff
@@ -0,0 +1,57 @@
+ROLLBACK TO hire_engelbart;
+COMMIT;
+
+SELECT name FROM researchers WHERE lab_id = 4;
```
I'd like to see some more tests. Some coverage for failure scenarios would be good. Maybe
- violate a constraint, see that we abort properly
- have an insert fail to a non-constraint reason (e.g. division by zero), see that ...
- check that we mark placements properly as aborted, likely by previously updating the placmenet's host, or by just dropping the table on the worker (via \c ...).

jasonmp85 replied:
Completely agreed. Will have an update soon. The "start a transaction and then do something on the worker that causes it to fail" is a little tougher here since (a) changing the connection endpoint during a test will kill the transaction, and (b) even if I could multiplex in some way, e.g. dropping the table and other operations could result in blocking (due to locks and the ongoing open transaction).

I'll push up some test before tomorrow and I'll ping you when they're up. Let me know if they're sufficient.

jasonmp85 replied:
Added quite a few tests, including `SELECT` after modify, DDL, `\copy`, primary key violations, and other types of errors. Used a `CONSTRAINT TRIGGER` to raise a non-constraint violation error type on a magic value. By making trigger deferred, could also test `COMMIT` failures.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -69,12 +87,63 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
+				uint64 *shardIdPtr = NULL;
+
+				MemSet(&participantKey, 0, sizeof(participantKey));
+				strncpy(participantKey.nodeName, placement->nodeName, MAX_NODE_LENGTH);
```
if you want to use a limited form of strcpy, it's generally better to use strlcpy, that guarantees to write a trailing 0. In this case the memset should guarantee the presence, but ...

jasonmp85 replied:
OK, I'll swap this (and also fix a similar occurrence in `connection_cache.c`).

jasonmp85 replied:
Oh, do you want me to check the return value/error on truncation, since that's a thing I can do with strlcpy?

marcocitus commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -718,6 +812,47 @@ ConsumeQueryResult(PGconn *connection, int64 *rows)
+
+			if (commit)
+			{
+				MarkParticipantShardsUnhealthy(participant);
```
Are we doing this even when there is only one placement?

anarazel replied:
@marcocitus We really shouldn't...

jasonmp85 replied:
Correct 😏

I caught onto this at some point but lost track of it when I refactored. It's not just for one placement, right now if ALL `COMMIT`s fail, we'll mark everything as inactive. That's pretty bad.

I'll add some logic here that's analogous to that in the modification itself, i.e. if all nodes fail to commit, don't mark any as unhealthy.

jasonmp85 replied:
Fixed this. Basically, if no node `COMMIT`s, we'll throw an error, rolling back any health modifications.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -110,6 +141,70 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
+
+		shardIdPtr = AllocateUint64(placement->shardId);
+
+		participantEntry->shardIds = list_make1(shardIdPtr);
```
I don't really care, but that's a lot of newlines ;)

jasonmp85 replied:
Removed one.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -751,3 +886,307 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+	}
+
+	/* reset transaction state */
+	hash_destroy(xactParticipantHash);
```
What about instead just unsetting the variable? Freeing the memory via context reset (automatic, at then end of xact) is a lot faster.

jasonmp85 replied:
BTW, I couldn't easily find a `hash_clear` thing that would erase hash contents without erasing the hash itself. Does such a thing exist? If so, I could keep the hash for the entire session (rather than recreating it for each top-level transaction).

jasonmp85 replied:
(If you think such an optimization is worthwhile; otherwise I'll just `NULL` out the variable like you ask).

anarazel replied:
> (If you think such an optimization is worthwhile; otherwise I'll just `NULL` out the variable like you ask).

just NULLing seems best.

jasonmp85 replied:
Done.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -751,3 +886,307 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+		XactParticipantEntry *participantEntry = NULL;
+		bool entryFound = false;
+
+		Assert(IsTransactionBlock());
```
That'll be wrong once the 'from function context' detection stuff is added.

jasonmp85 replied:
Yup.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -751,3 +886,307 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+	foreach(shardCell, participant->shardIds)
+	{
+		uint64 *shardIdPointer = (uint64 *) lfirst(shardCell);
+		uint64 shardId = (*shardIdPointer);
```
If only shard ids weren't pointlessly 64 bits.

jasonmp85 replied:
😈

anarazel commented on src/backend/distributed/utils/connection_cache.c:
```diff
@@ -152,42 +151,54 @@ PurgeConnection(PGconn *connection)
+	 * that case, we will want to close the provided connection in addition
+	 * to the one from the hash (which was closed by PurgeConnectionByKey).
+	 */
+	if (purgedConnection != connection)
```
Is that just paranoia (nothing against that), or is there a "valid" reason this can happen?

jasonmp85 replied:
I guess paranoia? It's been in the `connection_cache` since the `pg_shard` days, so didn't bother changing it now.

anarazel commented on src/backend/distributed/utils/datautils.c:
```diff
@@ -0,0 +1,32 @@
+uint64 *
+AllocateUint64(uint64 value)
+{
+	uint64 *allocatedValue = (uint64 *) palloc0(sizeof(uint64));
```
Not now, but it'd be imo better to use UInt64GetDatum/DatumGetUInt64 here. That "automatically" uses a pass-by-reference type if possible on the current platform, instead of incurring allocation overhead.

jasonmp85 replied:
(a) PostgreSQL doesn't have unsigned methods like that, no?

(b) I don't quite understand how this would help here… I _need_ to allocate a `uint64` in the `TopTransactionContext`, so it can be added to a list and live there in that context. If it were a smaller `int`, I could just put it in an `int` list, but there's nowhere to put a 64-bit value in a list, even it can be pass-by-value on a certain platform.

In other words, a pointer list is all I've got, so I need to point to something, which necessitates allocation. How can I circumvent allocation here? It seems less _overhead_ and more required. Am I missing something?

jasonmp85 replied:
> Not now, but […]

(And I understand I'm not changing anything here, just trying to understand what you mean, since it feels like I must be missing something crucial here)

anarazel replied:
On 2016-07-13 14:32:57 -0700, Jason Petersen wrote:
> > +
> > +#include "postgres.h"
> > +#include "c.h"
> > +
> > +#include "distributed/datautils.h"
> > +
> > +
> > +/*
> > + * Allocates a new uint64 on the heap and copies the input to that space,
> > + * returning a pointer to allocated memory. Useful for lists with 64-bit
> > + * integers.
> > + */
> > +uint64 *
> > +AllocateUint64(uint64 value)
> > +{
> > +	uint64 *allocatedValue = (uint64 *) palloc0(sizeof(uint64));
> 
> (a) PostgreSQL doesn't have unsigned methods like that, no?

It does:

/*
 * DatumGetUInt64
 *		Returns 64-bit unsigned integer value of a datum.
 *
 * Note: this macro hides whether int64 is pass by value or by reference.
 */

#ifdef USE_FLOAT8_BYVAL
#define DatumGetUInt64(X) ((uint64) GET_8_BYTES(X))
#else
#define DatumGetUInt64(X) (* ((uint64 *) DatumGetPointer(X)))
#endif
/*
 * UInt64GetDatum
 *		Returns datum representation for a 64-bit unsigned integer.
 *
 * Note: if int64 is pass by reference, this function returns a reference
 * to palloc'd space.
 */

#ifdef USE_FLOAT8_BYVAL
#define UInt64GetDatum(X) ((Datum) SET_8_BYTES(X))
#else
#define UInt64GetDatum(X) Int64GetDatum((int64) (X))
#endif

> (b) I don't quite understand how this would help here… I _need_ to allocate a `uint64` in the `TopTransactionContext`, so it can be added to a list and live there in that context. If it were a smaller `int`, I could just put it in an `int` list, but there's nowhere to put a 64-bit value in a list, even it can be pass-by-value on a certain platform.

You can just use a non-int list. Datums are guaranteed to be storable in
a size_t wide variable.

Andres

jasonmp85 replied:
> It does

I was searching the symbol index on my PostgreSQL `REL9_5_STABLE` checkout, which doesn't include this function (i.e. it's not yet in a released PostgreSQL). So we'd have to polyfill it anyhow.

> You can just use a non-int list. Datums are guaranteed to be storable in
a size_t wide variable.

Interesting. I was missing this piece of knowledge. Now I see how this could work, thanks!

marcocitus commented:
We should probably do something about non-transactional utility commands running inside a transaction:

    postgres=# BEGIN;
    BEGIN
    postgres=# INSERT INTO test VALUES(3);
    INSERT 0 1
    postgres=# ALTER TABLE test ADD COLUMN foo int;
    <wait forever>

The transaction performing the insert on the shard placement is holding on to a RowExclusiveLock, and alter table waits for an AccessExclusiveLock, so this creates a deadlock.

jasonmp85 commented:
Alright, I've pushed up fairly comprehensive tests (listed what they hit above) and fixed the remaining issues (the one @marcocitus recently raised, and the one around disallowing new connections during a modifying transaction).

I think that's all. One thing I _did_ notice is that I wanted to test router `SELECT`s (important) during modifying transactions, but I'm having a hard time triggering any. Even a simple select like `SELECT * FROM table WHERE id = 5` (partitioned by `id`) generates two tasks (there's a master merge task), which prevents the router executor from kicking in.

I just noticed this and I'm not sure which of the recent changes is keeping router `SELECT`s from going through, but as I've rebased this patch several times it could be several things (though neither `RETURNING` nor master-functions appear related. I'll look into this tomorrow and try to figure out what's going on.

@anarazel can you check this out in the morning and give a 👍 or 👎 about the tests and final changes? I'll flatten and merge after that (I hope).

jasonmp85 commented:
OK, so there remains a final bug here, which I only now recognize…

If you have two tables, each with one shard, (A and B) and two nodes (1 and 2), and table A is on _both_ nodes and table _B_ is only on node 2, the following can occur:

``sql
BEGIN;
INSERT INTO A VALUES (1);
INSERT INTO B VALUES (2);
COMMIT;
```

Because the first modification opens transactions to both nodes (1 and 2), it's valid to subsequently insert to B (which is in one of those participants). If that `INSERT` fails, we're fine, because we'd just abort the transaction (as the only active placement could not be modified).

But if all _statements_ succeed and we fail during the `COMMIT` phase, the following could happen:

  1. Send `COMMIT` to node 1 (succeeds)
  2. Send `COMMIT` to node 2 (fails)

Since node 2 failed, but _some_ node succeeded, we'll let the transaction succeed. But _no_ placement of table B has succeeded, so we have a problem. The transaction partially committed, and now the only placement of table B is unhealthy.

I think the only way around this (short of two-phase commit) is to say that subsequent modifications must hit _all_ of the nodes hit by an initial modification (i.e. no having colocated tables that happen to have lower replication).

Thoughts? Separate issue? Fix here?

anarazel commented:
Why do we need to mark B as unhealthy in the first place? There's no danger of divergence here?

anarazel commented on src/backend/distributed/executor/multi_client_executor.c:
```diff
@@ -94,6 +94,13 @@ MultiClientConnect(const char *nodeName, uint32 nodePort, const char *nodeDataba
+	if (IsModifyingTransaction)
+	{
+		ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+						errmsg("cannot open new connection during modifying "
```
I'm not entirely sure how, but it seems this message could be made a bit easier to understand for users.  Maybe something like "cannot open new connections in a transaction that has performed modifications", hint: After a modification has been executed only router-executable queries to exactly those shards are allowed. or something.

jasonmp85 replied:
Reworded it, but skipped that `HINT`, since they're supposed to be reserved for providing advice about how to solve a problem.

I suppose a more detailed description could go in the `detail`, but wanted to see if the new message was sufficient before adding that.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -44,7 +48,20 @@
+static HTAB *xactParticipantHash = NULL;
+static bool subXactAbortAttempted = false;
+static shmem_startup_hook_type prev_shmem_startup_hook = NULL;
 
```
IIRC we leave two newlines between vars and functions.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -791,3 +938,312 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+ */
+static void
+MarkParticipantShardsUnhealthy(XactParticipantEntry *participant)
+{
```
So I think this is a bit of the wrong logic.  We should be marking things as invalid based on the shards, not the transactions. I.e. track which shard have failed on which placements, and mark as invalid iff there more than one placement, and not all placements failed.

jasonmp85 commented:
It doesn't diverge, but the transaction is partially applied. The user
asked to do two things in a transaction, and only the first occurred.

I think the easiest behavior to explain is if you want to hit multiple
shards in a transaction, they need to be on all nodes in that transaction.
Thinking of just implementing that.

--
*Jason Petersen*
Software Engineer | Citus Data
303.736.9255
jason@citusdata.com

On Jul 15, 2016, at 6:45 PM, Andres Freund <notifications@github.com> wrote:

Why do we need to mark B as unhealthy in the first place? There's no danger
of divergence here?

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/citusdata/citus/pull/627#issuecomment-233098647>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AADF8fov6CGiJm0fkwDTWDjSLCgjJ-wfks5qWCmegaJpZM4JAaSV>
.

marcocitus commented:
> I think the easiest behavior to explain is if you want to hit multiple shards in a transaction, they need to be on all nodes in that transaction. Thinking of just implementing that. 

I think that's only easy until you run shard rebalancer. A single shard is the only thing guaranteed to be in one place.

jasonmp85 commented:
Well then, this code got a little uglier, but it now implements your alternate logic, @anarazel.

Now the transaction participant façade hash is of the same type as the connection cache one (it uses the same structs). Rather than putting shard identifiers on the connection objects within the hash, I'm making a map (well, list) from shard identifier to lists of connections that modified that shard.

A `NULL` connection entry in the cache or map means the connection had a problem somewhere along the way and the placement didn't make it to the end of the transaction.

So there's a walk across this map at the end of a successful transaction to ensure anything that failed during the `COMMIT` itself will be marked bad, but not if every connection failed for a given shard.

FWIW I find this immensely more confusing to explain as opposed to "if your shards aren't strictly colocated, you can't update multiple shards in a transaction", but it's an edge case at the end of the day, anyways.

jasonmp85 commented:
(Going to do a commenting pass in the morning; if you want to check out the logic before then, knock yourself out).

jasonmp85 commented on src/backend/distributed/master/master_metadata_utility.c:
```diff
@@ -125,19 +122,6 @@ LoadShardList(Oid relationId)
 }
 
 
-/* Allocates eight bytes, and copies given value's contents those bytes. */
```
Moving this back in the morning (not needed anymore).

jasonmp85 commented:
I'll also add a test for the case I mention above to verify your suggested behavior. Should be pretty easy with the `DEFERRED` `CONSTRAINT TRIGGER` hack I've got going.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -116,6 +149,67 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
+
+	xactParticipantHash = CreateXactParticipantHash();
+
+	oldContext = MemoryContextSwitchTo(TopTransactionContext);
```
Hm, are we actually allocating anything in here?

jasonmp85 replied:
Good catch. We _were_ when I had the `shardId` list on each hash entry (via `lappend`) but now are not. Removing.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -116,6 +149,67 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
+			}
+
+			PQclear(result);
+		}
```
Hm. Should we error out if we reach the connection == NULL case here?

jasonmp85 replied:
> Hm. Should we error out if we reach the `connection == NULL` case here?

No. A `NULL` connection just means that node has stopped being able to participate in the transaction, for whatever reason. A single node not being able to `BEGIN` isn't a reason to completely abort. I suppose if _all_ nodes could not start the transaction, we could abort here, but it's easier to just set all their connections to `NULL` and let the `Run` function error out (it will say "unable to modify any active placements").

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -44,7 +48,21 @@
+ * (not allowed) and a backend startup hook to register xact callbacks.
+ */
+static HTAB *xactParticipantHash = NULL;
+static List *xactShardConnSetList = NIL;
```
Should also be documented.

jasonmp85 replied:
Adding a paragraph about it.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -791,3 +930,341 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+
+
+static void
+MarkRemainingInactivePlacements()
```
(void)

jasonmp85 replied:
Fixing… :unamused:

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -791,3 +930,341 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+			}
+		}
+
+		if (successes == 0)
```
Needs some explanation.

jasonmp85 replied:
Done.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -791,3 +930,341 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+	MemoryContext oldContext = NULL;
+	List *connectionEntryList = NIL;
+
+	foreach(listCell, xactShardConnSetList)
```
For now that's fine, but in the longer run this probably needs to be a structure that's not O(N)

jasonmp85 replied:
Agreed. I initially created another hash, but it was quite onerous and given that this is _O(xact_shards * replication)_ it seems likely this will often run just as fast as any hash or tree.

anarazel commented:
Seems generally reasonable, mostly lacking comments.

jasonmp85 commented:
OK, @anarazel… anything else?

I'm going to do a pass on moving around my new functions to be in a more logical order (I have a tool I wrote that topologically sorts them, as a suggestion), but holding off on that until I've got the :shipit:.

Let me know if there's anything else to be done here.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -11,54 +11,106 @@
  */
 
-#include "postgres.h"
+#include "postgres.h" /* IWYU pragma: keep */
```
I still think that IWYU is one of the tools of the devil :P

jasonmp85 replied:
I know, I know! That said, if I'm relying on a data type in a header and don't include it myself, someone's going to deal with the breakage down the line when the transitive include is removed!

`#include`s are there for the machine; let the machine handle it, I say.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -11,54 +11,106 @@
+ * The following static variables are necessary to track the progression of
+ * multi-statement transactions managed by the router executor. After the first
+ * modification within a transaction (whether started by the use of a function
+ * or the use of BEGIN), the executor populates a hash with the transaction's
```
> (whether started by the use of a function or the use of BEGIN)

I think it's only BEGIN right now, IIRC we decided to punt on this initially.

jasonmp85 replied:
Fixed.

jasonmp85 commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -791,3 +1069,238 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+
+
+/*
+ * ExecuteTransactionEnd ends any remote transactions still taking place on
```
Noticed this comment is out of date. Fixing.

jasonmp85 replied:
Fixed.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -75,12 +127,21 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
 	Assert(task != NULL);
 
-	/* disallow transactions and triggers during distributed modify commands */
+	/* disallow triggers during distributed modify commands */
```
Unrelated to this commit: But I doubt that SKIP_TRIGGERS has any effect here? We don't ever reach nodeModifyTuple.c, which is responsible for firing triggers...

jasonmp85 replied:
I… don't remember. This was ported over from `pg_shard`, which may have been written with the assumption that we'll call whatever's next in the hook chain, which might well cause triggers to fire. Sorry, can't remember precisely what the deal was.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -75,12 +127,21 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
+		 * We could naturally handle function-based transactions (i.e. those
+		 * using PL/pgSQL or similar) by checking the type of queryDesc->dest,
+		 * but some customers already use functions that touch multiple shards
+		 * from within a function, so we'll ignore functions for now.
```
I think a followup issue about this would be good.

jasonmp85 replied:
Opened as #659.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -330,10 +481,11 @@ ExecuteTaskAndStoreResults(QueryDesc *queryDesc, Task *task,
 
 		ExecuteMasterEvaluableFunctions(query);
-		DeparseShardQuery(query, task, queryStringInfo);
+		deparse_shard_query(query, relid, task->anchorShardId, queryStringInfo);
```
Hm, not that I think it matters much, but that function had intentionally been introduced by @lithp. Why did you roll back that change?

jasonmp85 replied:
It was two lines to inline and was only used in exactly this one place; when reordering I didn't see a natural place for that helper function to live in this file, and it lacked a function comment.

I figured if it was simple enough to omit a comment and used exactly in one place I'd just inline it. Seemed weird to have to jump down several functions in the file to just see this little helper that just wrapped a call to `deparse_shard_query`.

anarazel replied:
Ok.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -510,6 +651,99 @@ ReturnRowsFromTuplestore(uint64 tupleCount, TupleDesc tupleDescriptor,
+ * GetConnectionForPlacement is the main entry point for acquiring a connection
+ * within the router executor. By using placements (rather than node names and
+ * ports) to identify connections, the router executor keep track of shards hit
+ * by a long-running transaction (and error out if a transaction attempts to
```
Hm, maybe s/long-running/multi-statement/? Or s/long-running/explicit/?

jasonmp85 replied:
Sure. Fixing.

jasonmp85 replied:
Done.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -791,3 +1068,238 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+		case XACT_EVENT_PREPARE:
+		case XACT_EVENT_PRE_PREPARE:
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
```
@marcocitus Hm, that's not a problem for this PR, but I wonder if there's a chance that this presents problems for MX?

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -791,3 +1068,238 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+RouterSubtransactionCallback(SubXactEvent event, SubTransactionId subId,
+							 SubTransactionId parentSubid, void *arg)
+{
+	if ((xactParticipantHash != NULL) && (event == SUBXACT_EVENT_ABORT_SUB))
```
What's the deal with the many parens here?

jasonmp85 replied:
Dunno, just grouping? There are several places in this file that do this.

anarazel replied:
Looks seriously weird to me; including the other places.

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -791,3 +1068,238 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+static void
+ExecuteTransactionEnd(bool commit)
+{
+	char *sqlCommand = commit ? "COMMIT TRANSACTION" : "ABORT TRANSACTION";
```
should be const.

jasonmp85 replied:
Done.

anarazel replied:
I guess you mean locally?

anarazel commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -791,3 +1068,238 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+		else
+		{
+			WarnRemoteError(connection, result);
+			participant->connection = NULL;
```
Hm. Is this connection ever going to be closed this way?

jasonmp85 replied:
Adding a call to `PurgeConnection`.

anarazel commented on src/include/distributed/multi_router_executor.h:
```diff
@@ -9,15 +9,37 @@
 #ifndef MULTI_ROUTER_EXECUTOR_H_
 #define MULTI_ROUTER_EXECUTOR_H_
 
+#include "c.h"
```
This really makes me decidedly unhappy.

anarazel commented on src/include/distributed/multi_router_executor.h:
```diff
@@ -9,15 +9,37 @@
+#include "nodes/pg_list.h"
+
+/* maximum (textual) lengths of hostname and port */
+#define MAX_NODE_LENGTH 255
```
How did we arrive at 255? IIRC that's roughly the length of the hostname limits, without the port itself. If we added 10 to that, we could handle all possible URL/port combinations.

jasonmp85 replied:
So… this comment is wrong. But also I don't need this `#define` anymore. Removing.

anarazel commented on src/test/regress/expected/multi_modifying_xacts.out:
```diff
@@ -0,0 +1,429 @@
+ 
+(1 row)
+
+SELECT master_create_distributed_table('labs', 'id', 'hash');
```
IIRC we also allow insert to append partitioned tables these days - so we probably should have a test for that here.

jasonmp85 replied:
Done. Added at bottom to avoid shard identifier churn.

anarazel commented on src/test/regress/expected/multi_modifying_xacts.out:
```diff
@@ -0,0 +1,429 @@
+-- creating savepoints should work...
+BEGIN;
+INSERT INTO researchers VALUES (5, 3, 'Dennis Ritchie');
+SAVEPOINT hire_thompson;
```
could you add a plpgsql check for that as well? Possibly just using a DO block?

jasonmp85 replied:
Done.

anarazel commented on src/test/regress/expected/multi_modifying_xacts.out:
```diff
@@ -0,0 +1,429 @@
+INSERT INTO labs VALUES (6, 'Bell Labs');
+\copy labs from stdin delimiter ','
+ERROR:  cannot open new connections after the first modification command within a transaction
+CONTEXT:  COPY labs, line 1: "7,E Corp"
```
A test doing it the other way round would be interesting, given that that opens connections to everything. Should afaics work correctly, but still would be nice to test.

anarazel commented on src/test/regress/expected/multi_modifying_xacts.out:
```diff
@@ -0,0 +1,429 @@
+WHERE  sp.shardid = s.shardid
+AND    s.logicalrelid = 'objects'::regclass;
+DELETE FROM objects;
+WARNING:  modified 0 tuples, but expected to modify 2
```
Uh, that's odd?

jasonmp85 replied:
Oh, right, yoou wanted a comment for this.

jasonmp85 replied:
Hm. If I just move the `DELETE` to before the "mark all as healthy" operation, I can avoid this error. Doing that.

jasonmp85 replied:
Done.

anarazel commented:
Barring resolution of the two test and some other minor issues this looks good to me.

sumedhpathak commented:
@jasonmp85 updates? Anything blocking you from getting this merged?

jasonmp85 commented:
@sumedhpathak four more tests, then I think nothing else:

pl/pgsql-generated SAVEPOINT
  • copy-first xact
  • copy-second, single-shard xact
  • (needed?) append-based table

jasonmp85 commented:
Turns out the "COPY after modification to a distributed table with only one shard" was already handled, so only the first, second, and last tests above were really needed. Added all. @anarazel, back to you.

marcocitus commented:
> copy-first xact
> copy-second, single-shard xact

There could be some problems with COPY in transactions. For example, you could get deadlocks on the advisory locks, since two transactions could lock shards in opposite order. For example, transaction A does COPY and locks shards 1-2-3, transaction B does INSERT on shard 3 and then a COPY and locks shard 1-2-3. They could get stuck waiting for each other, though I guess postgres would detect a deadlock in this case.

You could also get a distributed deadlock on the row-level locks, e.g. if you're writing the same primary key in multiple transactions, though this is a more general and much harder to solve problem.

jasonmp85 commented:
> In any case, COPY does not roll back after the statement finishes, so maybe we should just disallow it in transaction blocks.

That's what the latest DDL change did for DDLs. Adding a simple check to `COPY` could do the same and avoid all this reasoning about what is or is not allowed.

marcocitus commented:
> That's what the latest DDL change did for DDLs. Adding a simple check to COPY could do the same and avoid all this reasoning about what is or is not allowed.

Yeh, we added the check in DDL partly because of potential conflicts with single-shard transactions. I think adding the PreventTransactionChain thing to the COPY code path would make sense as well (e.g. in multi_utility.c, since the isTopLevel variable is available there).

jasonmp85 commented:
Got text ack from @anarazel, merging.

@marcocitus, I'll open a quick PR about preventing transactions in COPY and can discuss that there.

sumedhpathak commented:
👍!

On Thu, Jul 21, 2016 at 3:37 PM, Jason Petersen <notifications@github.com>
wrote:

> Merged #627 <https://github.com/citusdata/citus/pull/627>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/citusdata/citus/pull/627#event-731087071>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACeeC0r0Ve8O3wlyCy0EBOLzhpvDQ99cks5qX_S0gaJpZM4JAaSV>
> .
>



-- 
*Sumedh Pathak*
Citus Data
650.422.9797
sumedh@citusdata.com

jasonmp85 commented:
@marcocitus: we discussed this `COPY` concern (and DDLs) during stand up and I opened an issue, because we're not in agreement around whether we should ban its use in transactions, or what the alternative is… it's in #666.
